Sources/Signal.swift

```diff
@@ -237,9 +309,73 @@ public final class Signal<Value, Error: Swift.Error> {
 	}
 }
 
-private struct SignalState<Value, Error: Swift.Error> {
-	var observers: Bag<Signal<Value, Error>.Observer> = Bag()
-	var retainedSignal: Signal<Value, Error>?
+/// The status of a `Signal`.
+///
+/// - note: This three-case, multi-payload enum was verified to be a packed
+///         object reference on i386, x86-64, armv7 and armv8 under the Swift
+///         3.0.1 ABI. Should any changes be made to the enum, please first
+///         verify that the change would not cause the enum to spill over the
+///         size of an object reference on all platforms.
+private enum SignalStatus<Value, Error: Swift.Error> {
+	/// The `Signal` is alive.
+	case alive(SignalState<Value, Error, ()>)
+
+	/// The `Signal` has received a termination event, and is about to be
+	/// terminated.
+	case terminating(SignalState<Value, Error, Event<Value, Error>>)
+
+	/// The `Signal` has terminated.
+	case terminated
+}
+
+/// The state of a `Signal`.
+///
+/// As the amount of state would definitely span over a cache line,
+/// `SignalState` is set to be a reference type so that we can atomically
+/// update the reference instead.
+///
+/// - warning: In-place mutation should not be introduced to `SignalState`.
+///            Copy the states and create a new instance instead.
+private final class SignalState<Value, Error: Swift.Error, U> {
+	/// The observers of the `Signal`.
+	fileprivate let observers: Bag<Signal<Value, Error>.Observer>
+
+	/// A self-retaining reference. It is set when there are one or more active
+	/// observers.
+	fileprivate let retaining: Signal<Value, Error>?
+
+	fileprivate let associatedStorage: U?
```

Comment:
Relying on just `SignalState` is a significant regression, since now it has to load the reference and then load `terminationEvent` to check for termination in _all_ cases. The extra state provided by the enum does help here.

If you prefer to, I can revert 1681a44 and b53e796 and leave them to a separate PR.